spec-0007/index.md

```diff
@@ -5,69 +5,60 @@ author:
+- avoiding the use of global state and legacy bitstream generators, and
+- standardizing the usage and interpretation of an `rng` keyword for seeding.
 
-We suggest doing this by:
+We suggest implementing these principles by:
 
-- Passing the `rng` argument to `numpy.random.default_rng` to instantiate a `Generator`, and
-- deprecating the use of `np.seed` and `RandomState`.
+- deprecating the use of `RandomState` and `numpy.random.seed`,
+- accepting an `rng` argument in all functions that require random number generation, and
+- using `numpy.random.default_rng` to validate the `rng` argument and instantiate a `Generator`.
 
 ### Concepts
 
-- `BitGenerator`: Generates a stream of pseudo-random bits. The default generator in NumPy (`np.random.default_rng`) uses PCG64.
+- `BitGenerator`: Generates a stream of pseudo-random bits. The default generator in NumPy (`numpy.random.default_rng`) uses PCG64.
 - `Generator`: Derives pseudo-random numbers from the bits produced by a `BitGenerator`.
 - `RandomState`: a [legacy object in NumPy](https://numpy.org/doc/stable/reference/random/index.html), similar to `Generator`, that produces random numbers based on the Mersenne Twister.
 
 ### Constraints
 
 NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
 Common keyword arguments include `random_state` and `seed`.
-In practice, the seed is unfortunately also often controlled using `np.random.seed`.
+In practice, the seed is unfortunately also often controlled using `numpy.random.seed`.
 
 ## Implementation
 
-<!--
-Discuss how this would be implemented.
--->
-
 Legacy behavior in packages such as scikit-learn (`sklearn.utils.check_random_state`) typically handle `None` (use the global seed state), an int (convert to `RandomState`), or `RandomState` object.
 
 Two strong motivations for moving over to `Generator`s are:
 
 (1) they avoid naïve seeding strategies, such as using successive integers, via the underlying [SeedSequence](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning);
-(2) they avoid using global state (from `np.random.mtrand._rand`).
+(2) they avoid using global state (from `numpy.random.mtrand._rand`).
 
 Our recommendation here is a deprecation strategy which does not in _all_ cases adhere to the Hinsen[^hinsen] principle.
 
 The [deprecation strategy](https://github.com/scientific-python/specs/pull/180#issuecomment-1515248009) is:
 
-1. Accept both `rng` and `random_state` keyword arguments.
-2. If `rng=None`, handle `random_state` as in legacy behavior (see above), except use a compatible Generator instead of RandomState.
+1. Accept both `rng` and `random_state`/`seed` keyword arguments.
+2. If `rng=None`, handle `random_state`/`seed` as in legacy behavior (see above), except use a compatible Generator instead of RandomState.
    A DeprecationWarning is raised to warn about a future change in behavior.
```

Comment:
Should this section mention all the other things that happen during the deprecation period? The logic is a lot more involved than these four steps suggest.

Reply:
See the new documentation of the decorator for all the logic it implements.

Reply:
```suggestion
2. If both are specified, raise an error.
3. If neither is specified and `np.random.seed` has been used to set the seed, emit a `FutureWarning` about the upcoming change in behavior.
4. If `random_state`/`seed` is passed by keyword or by position, treat it as before, but:
  - Emit a `DeprecationWarning` if passed by keyword, warning about the deprecation of keyword `random_state` in favor of `rng`.
  - Emit a `FutureWarning` if passed by position, warning about the change in behavior of the positional argument.
5. If `rng` is passed by keyword, standardize it using `numpy.random.default_rng`.
```

Reply:
I'm a tad worried that the instructions are too detailed, causing confusion. Perhaps we can have a broad version, and then a more detailed discussion?

Reply:
Let's discuss it in real time. I think the implementation has been described in vague terms already, so at some point the details need to be laid out, and probably in English (like this) as well as the code.

Reply:
OK, that works for me, given that we've already given the high-level overview.